internal/ctrlutils/ingress-status.go

```diff
@@ -32,71 +33,96 @@ const (
-			// to use the reference IPs/Hosts from that LoadBalancer to update resource status addresses.
-			if err.Error() == proxyLBNotReadyMessage {
-				log.V(util.InfoLevel).Info("LoadBalancer type Service for the Kong proxy is not yet provisioned, waiting...", "service", publishService)
-				time.Sleep(time.Second)
-				continue
-			}
+type statusConfig struct {
+	ready             bool
+	cli               *clientset.Clientset
+	versionInfo       *version.Info
+	kubernetesVersion semver.Version
+	kiccli            *kicclientset.Clientset
+}
 
-			log.Error(err, "failed to determine kong proxy external ips/hostnames.")
-			return
-		}
-		break
-	}
+type statusInfo struct {
+	ready    bool
+	ips      []string
+	hostname string
+}
 
+func newStatusConfig(kubeConfig *rest.Config) (statusConfig, error) {
 	cli, err := clientset.NewForConfig(kubeConfig)
 	if err != nil {
-		log.Error(err, "failed to create k8s client.")
-		return
+		return statusConfig{ready: false}, err
 	}
 
 	versionInfo, err := cli.ServerVersion()
 	if err != nil {
-		log.Error(err, "failed to retrieve cluster version")
-		return
+		return statusConfig{ready: false}, err
 	}
 
 	kubernetesVersion, err := semver.Parse(strings.TrimPrefix(versionInfo.String(), "v"))
 	if err != nil {
-		log.Error(err, "could not parse cluster version")
-		return
+		return statusConfig{ready: false}, err
 	}
 
 	kiccli, err := kicclientset.NewForConfig(kubeConfig)
 	if err != nil {
-		log.Error(err, "failed to create kong ingress client.")
-		return
+		return statusConfig{ready: false}, err
 	}
+	return statusConfig{
+		ready:             true,
+		cli:               cli,
+		versionInfo:       versionInfo,
+		kubernetesVersion: kubernetesVersion,
+		kiccli:            kiccli,
+	}, nil
+}
 
+// PullConfigUpdate is a dedicated function that process ingress/customer resource status update after configuration is updated within kong.
+func PullConfigUpdate(
+	ctx context.Context,
+	kongConfig sendconfig.Kong,
+	log logr.Logger,
+	kubeConfig *rest.Config,
+	publishService string,
+	publishAddresses []string,
+) {
+	cfg := statusConfig{ready: false}
+	status := statusInfo{ready: false}
 	var wg sync.WaitGroup
+	var err error
 	for {
 		select {
 		case updateDone := <-kongConfig.ConfigDone:
-			log.V(util.DebugLevel).Info("data-plane updates completed, updating resource statuses")
-			wg.Add(1)
-			go func() {
-				if err := UpdateStatuses(ctx, &updateDone, log, cli, kiccli, &wg, ips, hostname, kubeConfig, kubernetesVersion); err != nil {
-					log.Error(err, "failed to update resource statuses")
+			if !cfg.ready {
+				cfg, err = newStatusConfig(kubeConfig)
+				if err != nil {
+					log.Error(err, "failed to initialize status updater")
 				}
-			}()
+			}
+			if !status.ready {
+				status, err = runningAddresses(ctx, kubeConfig, publishService, publishAddresses)
+				if err != nil {
+					if err.Error() == proxyLBNotReadyMessage {
```

Comment:
There is a pattern for well-defined errors:

https://pkg.go.dev/io#pkg-variables

> callers will test for EOF using ==

Define a global exported `var` with the error, and compare it directly instead of operating on the error message when constructing and testing for equality.

Reply:
c0e7a7fa though not exported; no exported function returns this and none should.